packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -5238,16 +5238,33 @@ class InheritedElement extends ProxyElement {
   InheritedElement(InheritedWidget super.widget);
 
   final Map<Element, Object?> _dependents = HashMap<Element, Object?>();
+  final Map<Type, InheritedElement> _lookupTable = HashMap<Type, InheritedElement>();
+
+  @pragma('vm:prefer-inline')
+  void _updateLookupTable() {
+    if (_lookupTable.isNotEmpty) {
+      return;
+    }
+    _lookupTable[widget.runtimeType] = this;
+    InheritedElement? ancestor = _inheritedAncestor;
+    while (ancestor != null) {
+      // Reading out the ancestor's lookup table is slower than just running
+      // up the tree based on local experiments.
+      _lookupTable.putIfAbsent(ancestor.widget.runtimeType, () => ancestor!);
+      ancestor = ancestor._inheritedAncestor;
+    }
+  }
+
+  @pragma('vm:prefer-inline')
+  InheritedElement? _lookup<T extends InheritedWidget>() {
+    _updateLookupTable();
+    return _lookupTable[T];
+  }
 
   @override
-  void _updateInheritance() {
-    assert(_lifecycleState == _ElementLifecycle.active);
-    final Map<Type, InheritedElement>? incomingWidgets = _parent?._inheritedWidgets;
-    if (incomingWidgets != null)
-      _inheritedWidgets = HashMap<Type, InheritedElement>.of(incomingWidgets);
-    else
-      _inheritedWidgets = HashMap<Type, InheritedElement>();
-    _inheritedWidgets![widget.runtimeType] = this;
+  void deactivate() {
+    _lookupTable.clear();
```

Comment:
In local tests, `if (_lookupTable.isEmpty) print('${widget.runtimeType} was not touched');` fires more than I would have expected for new gallery.

Reply:
question: does this happen for inherited elements that are FocuNodes or ActionMarkers?

Reply:
If I add a print here, run gallery, and tap material, I see:

```
flutter: DefaultTextStyle
flutter: _ParentInkResponseProvider
flutter: _ActionsMarker
flutter: _FocusMarker
flutter: DefaultTextStyle
flutter: _ParentInkResponseProvider
flutter: _ActionsMarker
flutter: _FocusMarker
```

Reply:
Tapping around and scrolling a list:

```
flutter: DefaultTextStyle
flutter: MediaQuery
flutter: IconTheme
flutter: DefaultTextStyle
flutter: DefaultTextStyle
flutter: DefaultTextStyle
flutter: _ParentInkResponseProvider
flutter: _ActionsMarker
flutter: _FocusMarker
flutter: IconTheme
flutter: DefaultTextStyle
flutter: _ParentInkResponseProvider
flutter: _ActionsMarker
flutter: _FocusMarker
flutter: IconTheme
flutter: _ParentInkResponseProvider
flutter: _ActionsMarker
flutter: _FocusMarker
flutter: MediaQuery
flutter: IconTheme
flutter: DefaultTextStyle
flutter: _ParentInkResponseProvider
flutter: _ActionsMarker
flutter: _FocusMarker
flutter: MediaQuery
flutter: IconTheme
flutter: DefaultTextStyle
flutter: _ParentInkResponseProvider
flutter: _ActionsMarker
flutter: _FocusMarker
flutter: MediaQuery
flutter: IconTheme
flutter: DefaultTextStyle
```

Reply:
This was wrong. The actions marker and focus markers are often looked up from what I can tell. When I properly guard the print the actions marker only shows up once as not getting looked up.

Reply:
_ParentInkResponseProvider on the other hand shows up frequently as not accessed.

Reply:
I wonder if we could get a similar effect by allowing only certain inherited elements to be lazy. _ActionsMarker/_FocusMarker won't be used at all if the user isn't interacting with the focus system or a kb

Reply:
ahh OK. Still, that is a lot of inherited widget copying for a a single type of widget. I still think the partitioning idea might be onto something...

Reply:
Clearing the list here is causing problems for FocusScope, which wants to do a lookup when it unfocuses...?

Reply:
its an inherited notifer